File: apps/eval-service/src/execution/planner_client.py
"""Real HTTP client to intelligence-service's PlannerService (HARD-7b/7h).

PlannerService is defined as a gRPC contract in alter/planner/v1/planner.proto,
but intelligence-service's actual transport is HTTP/FastAPI mirroring the RPC
names 1:1 (see apps/intelligence-service/src/planner/router.py's own module
doc). This client talks to that real HTTP surface, not a mocked one.
"""

from __future__ import annotations

import json
from dataclasses import dataclass

import httpx

DEFAULT_TIMEOUT_SECONDS = 30.0


@dataclass(frozen=True)
class SelectStrategyResult:
    strategy: str
    reason: str


@dataclass(frozen=True)
class DecomposeResult:
    stage_count: int
    entry_point: str
    stages: list[str]
    ambiguity_detected: bool


class PlannerClient:
    def __init__(
        self,
        base_url: str,
        timeout_seconds: float = DEFAULT_TIMEOUT_SECONDS,
        *,
        service_token: str = "",
    ) -> None:
        # Same optional-credential convention as IngestionEvalClient: attach
        # the internal service credential when one is configured so the
        # callee's app-level service auth accepts the call.
        headers = {"Authorization": f"Bearer {service_token}"} if service_token else None
        self._client = httpx.Client(base_url=base_url, timeout=timeout_seconds, headers=headers)

    def select_strategy(self, *, tenant_id: str, objective: str, mode: str) -> SelectStrategyResult:
        response = self._client.post(
            "/planner/select-strategy",
            json={"tenant_id": tenant_id, "objective": objective, "mode": mode},
        )
        response.raise_for_status()
        body = response.json()
        return SelectStrategyResult(strategy=body["strategy"], reason=body["reason"])

    def decompose(
        self,
        *,
        tenant_id: str,
        workspace_id: str,
        run_id: str,
        problem_spec_json: str,
        strategy: str,
    ) -> DecomposeResult:
        response = self._client.post(
            "/planner/decompose",
            json={
                "tenant_id": tenant_id,
                "workspace_id": workspace_id,
                "run_id": run_id,
                "strategy": strategy,
                "problem_spec_json": problem_spec_json,
            },
        )
        response.raise_for_status()
        body = response.json()
        skeleton = json.loads(body["task_skeleton_json"])
        stages = [node["key"].removeprefix("node_") for node in skeleton["nodes"]]
        return DecomposeResult(
            stage_count=len(skeleton["nodes"]),
            entry_point=skeleton["entry_point"],
            stages=stages,
            ambiguity_detected=body["ambiguity_detected"],
        )

    def understand(
        self, *, tenant_id: str, workspace_id: str, run_id: str, objective: str
    ) -> str:
        response = self._client.post(
            "/internal/problem-understanding/understand",
            json={
                "tenant_id": tenant_id,
                "workspace_id": workspace_id,
                "run_id": run_id,
                "objective": objective,
            },
        )
        response.raise_for_status()
        return json.dumps(response.json(), separators=(",", ":"))

    def close(self) -> None:
        self._client.close()

File: packages/adapters/src/http/planner-client.ts
// HTTP client for intelligence-service's PlannerService (PLAN-3). That
// service is exposed as FastAPI routes mirroring alter.planner.v1's RPC
// names (POST /planner/decompose, /replan, /select-strategy), not gRPC --
// see apps/intelligence-service/src/planner/router.py. Mirrors the
// Presidio/Tavily adapters' injectable-fetch pattern rather than
// inventing a new HTTP-calling convention.

import type {
  PlannerProblemContextReference as ProblemContextReference,
  PlannerProblemSpec,
} from "@alterx/contracts";

/** HTTP preserves the explicit JSON null used by the existing FastAPI route. */
export type ProblemSpec = Omit<PlannerProblemSpec, "current_situation"> & {
  readonly current_situation: string | null;
};
export type { ProblemContextReference };

export interface PlannerHttpClient {
  postJson(url: string, body: unknown): Promise<unknown>;
}

export function createFetchPlannerHttpClient(
  readServiceToken: () => string = defaultServiceToken,
): PlannerHttpClient {
  return {
    async postJson(url: string, body: unknown): Promise<unknown> {
      const response = await fetch(url, {
        method: "POST",
        headers: {
          "content-type": "application/json",
          authorization: `Bearer ${readServiceToken()}`,
        },
        body: JSON.stringify(body),
      });
      if (!response.ok) {
        throw new Error(
          `Planner request to ${url} failed with status ${response.status}`,
        );
      }
      return response.json();
    },
  };
}

export interface PlannerClientConfig {
  readonly baseUrl: string;
}

export interface DecomposeRequest {
  readonly tenant_id: string;
  readonly workspace_id: string;
  readonly run_id: string;
  readonly strategy: string;
  readonly problem_spec_json: string;
}

export interface ProblemUnderstandingRequest {
  readonly tenant_id: string;
  readonly workspace_id: string;
  readonly run_id: string;
  readonly objective: string;
  readonly actor_context?: Readonly<Record<string, string>>;
}

export interface DecomposeResponse {
  readonly task_skeleton_json: string;
  readonly ambiguity_detected: boolean;
  readonly clarification_questions: readonly string[];
}

export interface SelectStrategyRequest {
  readonly tenant_id: string;
  readonly objective: string;
  readonly mode: string;
}

export interface SelectStrategyResponse {
  readonly strategy: string;
  readonly reason: string;
}

export interface ReplanRequest {
  readonly tenant_id: string;
  readonly run_id: string;
  readonly current_dag_json: string;
  readonly failure_context_json: string;
}

export interface ReplanResponse {
  readonly revised_skeleton_json: string;
  readonly reason: string;
}

export interface PreparedCompilerInput {
  readonly status: "ready" | "blocked";
  readonly architecture?: unknown;
  readonly binding_decision?: unknown;
}

export interface PlannerHandler {
  understand(request: ProblemUnderstandingRequest): Promise<ProblemSpec>;
  decompose(request: DecomposeRequest): Promise<DecomposeResponse>;
  selectStrategy(request: SelectStrategyRequest): Promise<SelectStrategyResponse>;
  replan(request: ReplanRequest): Promise<ReplanResponse>;
  prepareCompilerInput?(request: { readonly tenant_id: string; readonly workspace_id: string; readonly task_skeleton: unknown }): Promise<PreparedCompilerInput>;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

export class PlannerResponseValidationError extends Error {
  constructor(operation: string, reason: string) {
    super(`Planner ${operation} response is invalid: ${reason}`);
    this.name = "PlannerResponseValidationError";
  }
}

function parseDecomposeResponse(raw: unknown): DecomposeResponse {
  if (
    !isRecord(raw) ||
    typeof raw.task_skeleton_json !== "string" ||
    typeof raw.ambiguity_detected !== "boolean" ||
    !Array.isArray(raw.clarification_questions) ||
    !raw.clarification_questions.every((q) => typeof q === "string")
  ) {
    throw new PlannerResponseValidationError("decompose", "missing or mistyped fields");
  }
  return {
    task_skeleton_json: raw.task_skeleton_json,
    ambiguity_detected: raw.ambiguity_detected,
    clarification_questions: raw.clarification_questions as readonly string[],
  };
}

function isStringArray(value: unknown): value is readonly string[] {
  return Array.isArray(value) && value.every((item) => typeof item === "string");
}

function parseProblemSpec(raw: unknown): ProblemSpec {
  if (
    !isRecord(raw) ||
    typeof raw.objective !== "string" ||
    !(raw.current_situation === null || typeof raw.current_situation === "string") ||
    !isStringArray(raw.actors) ||
    !isStringArray(raw.systems_involved) ||
    !isStringArray(raw.constraints) ||
    !isStringArray(raw.required_data) ||
    typeof raw.risk !== "string" ||
    !isStringArray(raw.missing_information) ||
    !isStringArray(raw.success_criteria) ||
    !Array.isArray(raw.context_references)
  ) {
    throw new PlannerResponseValidationError("understand", "missing or mistyped fields");
  }
  return raw as unknown as ProblemSpec;
}

function parseSelectStrategyResponse(raw: unknown): SelectStrategyResponse {
  if (
    !isRecord(raw) ||
    typeof raw.strategy !== "string" ||
    typeof raw.reason !== "string"
  ) {
    throw new PlannerResponseValidationError("select_strategy", "missing or mistyped fields");
  }
  return { strategy: raw.strategy, reason: raw.reason };
}

function parseReplanResponse(raw: unknown): ReplanResponse {
  if (
    !isRecord(raw) ||
    typeof raw.revised_skeleton_json !== "string" ||
    typeof raw.reason !== "string"
  ) {
    throw new PlannerResponseValidationError("replan", "missing or mistyped fields");
  }
  return { revised_skeleton_json: raw.revised_skeleton_json, reason: raw.reason };
}

export class PlannerClient implements PlannerHandler {
  constructor(
    private readonly config: PlannerClientConfig,
    private readonly httpClient: PlannerHttpClient = createFetchPlannerHttpClient(),
  ) {}

  async understand(request: ProblemUnderstandingRequest): Promise<ProblemSpec> {
    const raw = await this.httpClient.postJson(
      `${this.config.baseUrl}/internal/problem-understanding/understand`,
      request,
    );
    return parseProblemSpec(raw);
  }

  async decompose(request: DecomposeRequest): Promise<DecomposeResponse> {
    const raw = await this.httpClient.postJson(
      `${this.config.baseUrl}/planner/decompose`,
      request,
    );
    return parseDecomposeResponse(raw);
  }

  async selectStrategy(
    request: SelectStrategyRequest,
  ): Promise<SelectStrategyResponse> {
    const raw = await this.httpClient.postJson(
      `${this.config.baseUrl}/planner/select-strategy`,
      request,
    );
    return parseSelectStrategyResponse(raw);
  }

  async replan(request: ReplanRequest): Promise<ReplanResponse> {
    const raw = await this.httpClient.postJson(
      `${this.config.baseUrl}/planner/replan`,
      request,
    );
    return parseReplanResponse(raw);
  }

  async prepareCompilerInput(request: { readonly tenant_id: string; readonly workspace_id: string; readonly task_skeleton: unknown }): Promise<PreparedCompilerInput> {
    const raw = await this.httpClient.postJson(`${this.config.baseUrl}/internal/architecture-synthesis/prepare-compiler-input`, request);
    if (!isRecord(raw) || (raw.status !== "ready" && raw.status !== "blocked")) throw new PlannerResponseValidationError("prepare_compiler_input", "invalid response");
    if (raw.status === "ready" && (raw.architecture === undefined || raw.binding_decision === undefined)) throw new PlannerResponseValidationError("prepare_compiler_input", "missing architecture binding");
    return raw as unknown as PreparedCompilerInput;
  }
}

/**
 * Same fail-closed real-credential convention as SelectionBindingClient's
 * defaultServiceToken -- a missing INTERNAL_SERVICE_TOKEN must not degrade
 * into an anonymous request the callee then has to decide about.
 */
function defaultServiceToken(): string {
  const token = process.env["INTERNAL_SERVICE_TOKEN"]?.trim();
  if (!token) {
    throw new Error(
      "INTERNAL_SERVICE_TOKEN is required to call Planner",
    );
  }
  return token;
}
